node.go

```diff
@@ -358,14 +403,26 @@ func (n *ProtoNode) Multihash() mh.Multihash {
 	return n.cached.Hash()
 }
 
-// Links returns the node links.
+// Links returns a copy of the node's links.
 func (n *ProtoNode) Links() []*format.Link {
-	return n.links
+	if n.linksDirty {
+		// there was a mutation involving links, make sure we sort
+		sort.Stable(LinkSlice(n.links))
+		n.linksDirty = false
+		n.encoded = nil
+	}
+	links := make([]*format.Link, len(n.links))
+	copy(links, n.links)
+	return links
```

Comment:
```suggestion
	return append([]*format.link(nil), n.links...)
```
This is the canonical way to make a slice copy because `make` will set the capacity to the len precisely, while `append` while extend the capacity to the underlying allocation size.
See `slices.Clone` https://pkg.go.dev/golang.org/x/exp/slices#Clone
```go
func Clone[S ~[]E, E any](s S) S {
	// Preserve nil in case it matters.
	if s == nil {
		return nil
	}
	return append(S([]E{}), s...)
}
```
You can also bump this package to `go1.18` and import `golang.org/x/exp/slices` if you want.

Reply:
👍 I didn't know this, but also I don't think I understand the nuance here. Is the problem that by setting the length I'm also setting the capacity, so I'm imposing a risk of a full reallocation on the user if they want to modify it themselves, whereas using `append` like this will use a default block allocation for the slice so I'm giving breathing room for modification?

Reply:
Yes exactly.

The go allocator often sligthly round up, with make this extra space do nothing.
While with append it add to the capacity of the slice and can sometime save an allocation if this slice is appended to.